src/player/process-death.cpp

```diff
@@ -42,172 +42,149 @@
  * Centers a string within a GRAVE_LINE_WIDTH character string		-JWT-
  * @details
  */
-static void center_string(char *buf, concptr str)
+static std::string center_string(const std::string &str)
 {
-    int i = strlen(str);
-    int j = GRAVE_LINE_WIDTH / 2 - i / 2;
-    (void)sprintf(buf, "%*s%s%*s", j, "", str, GRAVE_LINE_WIDTH - i - j, "");
+    int j = GRAVE_LINE_WIDTH / 2 - str.length() / 2;
+    return std::string().append(j, ' ').append(str).append(GRAVE_LINE_WIDTH - str.length() - j, ' ');
 }
 
 /*!
  * @brief 墓に基本情報を表示
  * @param player_ptr プレイヤーへの参照ポインタ
- * @param buf 墓テンプレ
  */
-static void show_basic_params(PlayerType *player_ptr, char *buf)
+static void show_basic_params(PlayerType *player_ptr)
 {
-    char tomb_message[160];
-    (void)sprintf(tomb_message, _("レベル: %d", "Level: %d"), (int)player_ptr->lev);
-    center_string(buf, tomb_message);
-    put_str(buf, 11, 11);
-
-    (void)sprintf(tomb_message, _("経験値: %ld", "Exp: %ld"), (long)player_ptr->exp);
-    center_string(buf, tomb_message);
-    put_str(buf, 12, 11);
-
-    (void)sprintf(tomb_message, _("所持金: %ld", "AU: %ld"), (long)player_ptr->au);
-    center_string(buf, tomb_message);
-    put_str(buf, 13, 11);
+    put_str(center_string(format(_("レベル: %d", "Level: %d"), (int)player_ptr->lev)).data(), 11, 11);
+
+    put_str(center_string(format(_("経験値: %ld", "Exp: %ld"), (long)player_ptr->exp)).data(), 12, 11);
+
+    put_str(center_string(format(_("所持金: %ld", "AU: %ld"), (long)player_ptr->au)).data(), 13, 11);
 }
 
 #ifdef JP
 /*!
  * @brief プレイヤーを殺したモンスターを表示する (日本語版専用)
  * @param player_ptr プレイヤーへの参照ポインタ
- * @param buf 墓テンプレ
- * @param tomb_message 墓碑に刻む言葉
- * @return 追加の行数
+ * @return メッセージと余分な行数のペア
  */
-static int show_killing_monster(PlayerType *player_ptr, char *buf, char *tomb_message, size_t tomb_message_size)
+static std::pair<std::string, int> show_killing_monster(PlayerType *player_ptr)
 {
-    shape_buffer(player_ptr->died_from.data(), GRAVE_LINE_WIDTH + 1, tomb_message, tomb_message_size);
-    char *t;
-    t = tomb_message + strlen(tomb_message) + 1;
+    char buf[160];
+    shape_buffer(player_ptr->died_from.data(), GRAVE_LINE_WIDTH + 1, buf, sizeof(buf));
+    char *t = buf + strlen(buf) + 1;
     if (!*t) {
-        return 0;
+        return std::make_pair(buf, 0);
     }
 
-    char killer[MAX_MONSTER_NAME];
-    strcpy(killer, t); /* 2nd line */
+    std::string killer = t; /* 2nd line */
     if (*(t + strlen(t) + 1)) /* Does 3rd line exist? */
     {
-        for (t = killer + strlen(killer) - 2; iskanji(*(t - 1)); t--) { /* Loop */
-            ;
+        auto i = killer.length() - 2;
+        while (iskanji(killer[i - 1])) {
```

Comment:
If `killer.length()` is less than or equal to 2, an out-of-array reference is occurred.

This part of the code is originally suspect and will be refactored later, but for now, please also check that `i > 0` in the while statement as a countermeasure.

Reply:
The latest forced push adds an i > 0 to the while and adjusts the initial assignment of i when the length is less than 3 in case the type of length is unsigned.